### PR TITLE
CUMULUS-2606: associate an execution with a granule via API

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-2606: As a user, I want to associate an execution with a granule](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2606)

## Changes

* Adds `PUT /granules/{granuleId}/execution` API endpoint to associate an execution with a granule
* Adds helper `associateExecutionWithGranule` to `@cumulus/api-client/granules`
* Add integration test
* cumulus-api PR: https://github.com/nasa/cumulus-api/pull/311

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,7 +6,17 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
 
+### Added
+
+- **CUMULUS-2576**
```
Comment (kkelly51) [inline CHANGELOG.md:11]: 2576 didn't go out in the 9.5.0 release?

File: packages/api/endpoints/granules.js [hunk 2]
```diff
@@ -257,6 +257,67 @@ async function put(req, res) {
...
+ */
+const associateExecution = async (req, res) => {
+  const granuleName = req.params.granuleName;
+
+  const { collectionId, granuleId, executionArn } = req.body || {};
+  if (!granuleId || !collectionId || !executionArn) {
+    return res.boom.badRequest('Field granuleId, collectionId or executionArn is missing from request body');
+  }
+
+  if (granuleName !== granuleId) {
```
Comment (kkelly51) [inline packages/api/endpoints/granules.js:275]: Is there a case where these would not equal each other? Doesn't the granuleId come from the request?

File: packages/api/endpoints/granules.js [hunk 3]
```diff
@@ -257,6 +257,67 @@ async function put(req, res) {
...
+    if (error instanceof RecordDoesNotExist) {
+      if (granule === undefined) {
+        return res.boom.notFound(`No granule found to associate execution with for granuleId ${granuleId}`);
+      }
+      return res.boom.notFound(`Execution ${executionArn} not found`);
+    }
+    return res.boom.badRequest(errorify(error));
+  }
+
+  if (granule.collectionId !== collectionId) {
```
Comment (kkelly51) [inline packages/api/endpoints/granules.js:300]: Why do we care about the collectionId? Don't we just need the granuleId to update the granule?

**Review by kkelly51 — COMMENTED** [review]
Leaving comments so you can see them, about to test now

File: CHANGELOG.md (hunk 1 above)
Comment (jennyhliu) [reply CHANGELOG.md:11]: 2576 is not in 9.5.0

File: packages/api/endpoints/granules.js (hunk 2 above)
Comment (jennyhliu) [reply packages/api/endpoints/granules.js:275]: We update a single granule, and want to make sure the granuleId in the request matches the one in the path.

File: packages/api/endpoints/granules.js (hunk 3 above)
Comment (jennyhliu) [reply packages/api/endpoints/granules.js:300]: In postgres schema, granuleId + collectionId is unique, collectionId will be part of the query for getting the granule when dynamoDb goes away.

**Review by kkelly51 — APPROVED** [review]
Checks out

**Pull request merged**